cachi2/core/models/input.py

```diff
@@ -121,30 +119,32 @@ class Request(pydantic.BaseModel):
     packages: list[PackageInput]
     flags: frozenset[Flag] = frozenset()
 
-    @pydantic.validator("packages")
+    @pydantic.field_validator("packages")
     def _unique_packages(cls, packages: list[PackageInput]) -> list[PackageInput]:
         """De-duplicate the packages to be processed."""
         return unique(packages, by=lambda pkg: (pkg.type, pkg.path))
 
-    @pydantic.validator("packages", each_item=True)
-    def _check_package_paths(cls, package: PackageInput, values: dict) -> PackageInput:
+    @pydantic.field_validator("packages")
+    def _check_packages_paths(
+        cls, packages: list[PackageInput], info: pydantic.ValidationInfo
```

Comment:
Alternatively, we could use [pydantic.functional_validators.AfterValidator](https://docs.pydantic.dev/latest/concepts/validators/#generic-validated-collections) to still apply the validation for `each_item`. What I don't like about it is that we'd need to move this function to outside the class, diverging from the rest of validator functions. It does look better without the loop, on the other hand. So take this as a nitpick/suggestion.

Here's how I tested it locally:
```
diff --git a/cachi2/core/models/input.py b/cachi2/core/models/input.py
index 198a158..84d866b 100644
--- a/cachi2/core/models/input.py
+++ b/cachi2/core/models/input.py
@@ -7,6 +7,8 @@ from cachi2.core.errors import InvalidInput
 from cachi2.core.models.validators import check_sane_relpath, unique
 from cachi2.core.rooted_path import PathOutsideRoot, RootedPath
 
+from pydantic.functional_validators import AfterValidator
+
 if TYPE_CHECKING:
     from pydantic.error_wrappers import ErrorDict
 
@@ -111,12 +113,29 @@ PackageInput = Annotated[
 ]
 
 
+def _check_packages_paths(package: PackageInput, info: pydantic.ValidationInfo) -> list[PackageInput]:
+    """Check that package paths are existing subdirectories."""
+    source_dir: RootedPath = info.data.get("source_dir", None)
+    if source_dir is not None:
+        try:
+            abspath = source_dir.join_within_root(package.path)
+        except PathOutsideRoot:
+            raise ValueError(
+                f"package path (a symlink?) leads outside source directory: {package.path}"
+            )
+        if not abspath.path.is_dir():
+            raise ValueError(
+                f"package path does not exist (or is not a directory): {package.path}"
+            )
+    return package
+
+
 class Request(pydantic.BaseModel):
     """Holds all data needed for the processing of a single request."""
 
     source_dir: RootedPath
     output_dir: RootedPath
-    packages: list[PackageInput]
+    packages: list[Annotated[PackageInput, AfterValidator(_check_packages_paths)]]
     flags: frozenset[Flag] = frozenset()
 
     @pydantic.field_validator("packages")
@@ -124,26 +143,6 @@ class Request(pydantic.BaseModel):
         """De-duplicate the packages to be processed."""
         return unique(packages, by=lambda pkg: (pkg.type, pkg.path))
 
-    @pydantic.field_validator("packages")
-    def _check_packages_paths(
-        cls, packages: list[PackageInput], info: pydantic.ValidationInfo
-    ) -> list[PackageInput]:
-        """Check that package paths are existing subdirectories."""
-        source_dir: RootedPath = info.data.get("source_dir", None)
-        if source_dir is not None:
-            for p in packages:
-                try:
-                    abspath = source_dir.join_within_root(p.path)
-                except PathOutsideRoot:
-                    raise ValueError(
-                        f"package path (a symlink?) leads outside source directory: {p.path}"
-                    )
-                if not abspath.path.is_dir():
-                    raise ValueError(
-                        f"package path does not exist (or is not a directory): {p.path}"
-                    )
-        return packages
-
     @pydantic.field_validator("packages")
     def _packages_not_empty(cls, packages: list[PackageInput]) -> list[PackageInput]:
         """Check that the packages list is not empty."""
```